// pdf-invoice-hu/dist/i18n.js
'use strict';

module.exports = {
  	en_US: {
		Invoice: 'Invoice',
		createdAt: 'Created',
		completionDate: 'Completion date',
		paymentDeadline: 'Payment deadline',
		paymentMethod: 'Payment method',
		invoiceId: 'Invoice No.',
		amount: 'Total',
		chargeStatus: 'Billing status',
		chargeFor: 'Billing for: ',
		description: 'Description',
		name: 'Name',
		quantity: 'Quantity',
		seller: 'Seller',
		buyer: 'Buyer',
		address: 'Address',
		vat: 'VAT number',
		eu_vat: 'EU VAT',
		phone: 'Phone',
		email: 'E-mail',
		bank_acc: 'Bank Acc.',
		unit_price: 'Unit price',
		net_price: 'Net price',
		vat_amount: 'VAT',
		total: 'Total',
		totalNetAmount: 'Total net amount',
		totalGrossAmount: 'Total gross amount',
		totalVatAmount: 'Total VAT amount',
		forint: 'HUF',
		madeWith: '{ Made with DDO DreamInvoice Billing }',
  	},
  	hu_HU: {
		Invoice: 'Számla',
		createdAt: 'Számla kelte',
		completionDate: 'Teljesítési határidő',
		paymentDeadline: 'Fizetési határidő',
		paymentMethod: 'Fizetés módja',
		invoiceId: 'Számla száma',
		amount: 'Bruttó ár',
		chargeStatus: 'Számla típusa',
		chargeFor: 'Vásárló: ',
		description: 'Leírás',
		name: 'Megnevezés',
		quantity: 'Mennyiség',
		seller: 'Eladó',
		buyer: 'Vevő',
		address: 'Cím',
		vat: 'Adószám',
		eu_vat: 'EU Adószám',
		phone: 'Telefon',
		email: 'E-mail',
		bank_acc: 'Bankszámlaszám',
		unit_price: 'Egységár',
		net_price: 'Netto',
		vat_amount: 'Áfa',
		total: 'Összesen',
		totalNetAmount: 'Nettó összeg',
		totalGrossAmount: 'Bruttó összeg',
		totalVatAmount: 'ÁFA összesen',
		forint: 'Ft',
		madeWith: '{ Készült a DDO DreamInvoice számlázóval }',
	}
};
